File: src/main/kotlin/uk/ac/ic/benchmatte/util/Benchmark.kt
package uk.ac.ic.benchmatte.util

import matting.OpenCVMatter
import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.imgcodecs.Imgcodecs
import uk.ac.ic.matterialize.matting.BackgroundNegationMatter
import uk.ac.ic.matterialize.matting.FaceDetectionMatter
import uk.ac.ic.matterialize.matting.KMeansMatter
import uk.ac.ic.matterialize.matting.Matter
import uk.ac.ic.matterialize.matting.MatterMode
import java.io.File
import uk.ac.ic.matterialize.matting.backgroundcut.BackgroundCutMatter

class Benchmark(private val masksPath: String, private val backgroundsPath: String, private val foregroundsPath: String) {
    val backgrounds = mutableListOf<Mat>()
    val foregrounds = mutableListOf<Mat>()

    val composeds = mutableListOf<Mat>()
    val masks = mutableListOf<Mat>()

    private val matters = mutableListOf<Triple<MutableList<Pair<MatterMode, Matter>>, Mat, Mat>>()

    fun setup() {

        File(backgroundsPath).listFiles().zip(File(foregroundsPath).listFiles()).forEach { (bg, fg) ->

            println("Composing (bg) $bg with (fg) $fg")

            val bgMat = Compositor.loadFile(bg.path)
            val fgMat = Compositor.loadFile(fg.path)

            backgrounds.add(bgMat)
            foregrounds.add(fgMat)

            val pair = Compositor.compose(bgMat, fgMat, Compositor.loadFile(masksPath + "/" + fg.name, CvType.CV_8U))
            composeds.add(pair.first)
            masks.add(pair.second)

            val imgMatters = mutableListOf<Pair<MatterMode, Matter>>()

            MatterMode.values().forEach {
                val matter = when (it) {
                    MatterMode.KMeans -> KMeansMatter(bgMat)
                    MatterMode.FaceDetection -> FaceDetectionMatter()
                    MatterMode.BackgroundNegation -> BackgroundNegationMatter(bgMat)
                    MatterMode.OpenCV -> OpenCVMatter()
                    MatterMode.BackgroundCut -> BackgroundCutMatter(bgMat)
                }

                // the initial run is discarded to account for initialisation times
                matter.backgroundMask(pair.first)

                imgMatters.add(Pair(it, matter))
            }

            matters.add(Triple(imgMatters, pair.first, pair.second))
        }
    }

    fun export() {
        (0 until composeds.size).forEach { i ->
            Imgcodecs.imwrite("images/output/${i}_composed.png", composeds[i])
            Imgcodecs.imwrite("images/output/${i}_mask.png", masks[i])
        }
    }

    fun run(): List<List<Triple<MatterMode, Array<IntArray>, Long>>> {
        var i = 0
        return matters.map { (imgMatter, composed, mask) ->
            i++
            imgMatter.map { (mode, matter) ->
                val start = System.currentTimeMillis()
                val result = matter.backgroundMask(composed)
                val end = System.currentTimeMillis()

                Imgcodecs.imwrite("images/output/${i - 1}_$mode.png", result)

                val confusion = Scorer.difference(result, mask).first

                Triple(mode, confusion, end - start)
            }
        }
    }
}

File: src/benchmatte/benchmark.cpp
#include "benchmark.hpp"
#include "../matting/background_cut/background_cut_matter.hpp"
#include "../matting/background_negation_matter.hpp"
#include "../matting/opencv_matter.hpp"
#include "scorer.hpp"
#include <chrono>

using namespace chrono;

vector<BenchmarkResult> Benchmark::run() {
  vector<BenchmarkResult> result;
  int num_matters = 3;
  for (int i = 0; i < backgrounds.size(); i++) {
    OpenCVMatter opencv_matter = OpenCVMatter();
    BackgroundNegationMatter background_negation_matter =
        BackgroundNegationMatter(backgrounds[i]);
    BackgroundCutMatter background_cut_matter =
        BackgroundCutMatter(backgrounds[i]);
    IMatter *matters[] = {
        &opencv_matter, &background_negation_matter, &background_cut_matter};
    string names[] = {
        "OpenCVMatter", "BackgroundNegationMatter", "BackgroundCutMatter"};
    for (int j = 0; j < num_matters; j++) {
      auto start{system_clock::now()};
      Mat mask = matters[j]->background_mask(compositions[i]);
      long time =
          duration_cast<milliseconds>(system_clock::now() - start).count();
      imwrite(
          "../images/output/" + to_string(i) + "_" + names[j] +
              "_generated_mask.png",
          mask);
      ConfusionMatrix stats = Scorer::difference(mask, masks[i]);
      result.emplace_back(names[j], stats, time);
    }
    break;
  }
  return result;
}

void Benchmark::export_images() {
  for (int i = 0; i < compositions.size(); i++) {
    imwrite(
        "../images/output/" + to_string(i) + "_composed.png", compositions[i]);
    imwrite("../images/output/" + to_string(i) + "_mask.png", masks[i]);
  }
}
